toolchain/parse/tree.cpp

```diff
@@ -235,6 +246,15 @@ auto Tree::Verify() const -> ErrorOr<Success> {
       return Error(llvm::formatv(
           "Node #{0} is a placeholder node that wasn't replaced.", n.index));
     }
+    // Should extract successfully if node not marked as having an error.
+    if (!n_impl.has_error && !TestExtract(this, n, n_impl.kind, nullptr)) {
```

Comment:
Can you measure any performance impact from this additional verify step? (I don't know that we have a good benchmark here, but maybe you could generate a large input and run it through `carbon compile`.)